PCF-ADR-0004-codeflare-test-strategy.md

```diff
@@ -0,0 +1,376 @@
+  * Grant the MCAD controller ServiceAccount edit permission on RayCluster resources
+  * Create a default InstaScale resource
+Wait until all the components are ready
+The e2e test cases can now be executed, with a **standard user role**, ideally in parallel, to speed the execution time up and shorten the feedback loop as much as possible.
+> **Note**
+> The test cases should define lean / minimal compute resources, relative to the provisioned cluster, so parallel execution / throughput of the batch jobs submitted via the MCAD scheduling queue is maximized.
+
+##### Submit a Sample PyTorch Job in a managed Ray Cluster
+
+###### Description
+
+Submit a test PyTorch batch job, to a Ray cluster managed by MCAD, in a user tenant, and assert successful completion of the job. Shutdown the Ray cluster, and assert successful freeing of resources.
+
+###### Scenario
+
+1. Create a test Namespace (resp. Project)
+2. Create a test AppWrapper resource with the following specifications: https://github.com/project-codeflare/multi-cluster-app-dispatcher/blob/14d569bec1cd016dd41352e3c026f461d851f480/doc/usage/examples/kuberay/config/aw-raycluster.yaml
+3. Wait until the Ray cluster is ready
+4. Submit the test batch job to the Ray cluster
+5. Wait until the job has completed
+6. Assert the job status is successful
+7. Delete the test AppWrapper resource
+8. Assert all the resources have been successfully freed
+
+##### Submit a Sample PyTorch Job directly via MCAD
+
+Lower priority.
+
+###### Description
+
+Submit a test PyTorch batch job to the MCAD scheduler, in a user tenant, and assert successful completion of the job. Assert successful execution of the job, and clean-up of resources upon the batch job completion.
+
+###### Scenario
+
+1. Create a test Namespace (resp. Project)
+2. Create a test AppWrapper resource with the following specifications:
+``` 
+apiVersion: mcad.ibm.com/v1beta1
+kind: AppWrapper
+spec:
+ resources:
+   GenericItems:
+     - allocated: 0
+       generictemplate:
+         apiVersion: v1
+         kind: Pod
+         metadata:
+         spec:
+           containers:
+             - command:
+                 - bash
+                 - '-c'
+                 - >-
+                   torchrun --rdzv_backend c10d --rdzv_endpoint
+                   $TORCHX_RANK0_HOST:49782 --rdzv_id 'test-job'
+                   --nnodes 1 --nproc_per_node 1 --node_rank '0' --tee 3 --role
+                   '' mnist.py
+               env:
+                 - name: TORCHX_TRACKING_EXPERIMENT_NAME
+                   value: default-experiment
+                 - name: LOGLEVEL
+                   value: WARNING
+                 - name: TORCHX_JOB_ID
+                   value: 'kubernetes_mcad://torchx/test-job'
+                 - name: TORCHX_RANK0_HOST
+                   value: localhost
+                 - name: TORCHX_MCAD_MNIST_0_HOSTS
+                   value: test-job
+               image: 'quay.io/michaelclifford/mnist-test:latest'
```

Comment:
we probably want to move this image into the codeflare quay org.

Reply:
It won't be necessary, as we managed to rely on the _generic_ PyTorch image, and have the test inputs dynamically mounted / installed: https://github.com/project-codeflare/codeflare-operator/pull/136/files#diff-a2425bf13e5157c5c381d21407a576c5c3b5260dc338514ef67bbc63990d7b64R80-R87